crates/bitcoind_rpc/src/bip158.rs

```diff
@@ -171,27 +150,71 @@ impl<C: RpcApi> Iterator for FilterIter<'_, C> {
     type Item = Result<Event, Error>;
 
     fn next(&mut self) -> Option<Self::Item> {
-        (|| -> Result<_, Error> {
-            // if the next filter matches any of our watched spks, get the block
-            // and return it, inserting relevant block ids along the way
-            self.next_filter()?.map_or(Ok(None), |(block, filter)| {
-                let height = block.height;
-                let hash = block.hash;
-
-                if self.spks.is_empty() {
-                    Err(Error::NoScripts)
-                } else if filter
-                    .match_any(&hash, self.spks.iter().map(|script| script.as_bytes()))
-                    .map_err(Error::Bip158)?
-                {
-                    let block = self.client.get_block(&hash)?;
-                    self.blocks.insert(height, hash);
-                    let inner = EventInner { height, block };
-                    Ok(Some(Event::Block(inner)))
-                } else {
-                    Ok(Some(Event::NoMatch(height)))
+        (|| -> Result<Option<_>, Error> {
+            if self.height > self.stop {
+                return Ok(None);
+            }
+            // Fetch next header.
+            let mut height = self.height;
+            let mut hash = self.client.get_block_hash(height as u64)?;
+
+            let mut reorg_depth = 0;
+
+            let header = loop {
+                if reorg_depth >= Self::MAX_REORG_DEPTH {
+                    return Err(Error::ReorgDepthExceeded);
+                }
+
+                let header = self.client.get_block_header(&hash)?;
+
+                let prev_height = height.saturating_sub(1);
+                match self.headers.get(&prev_height).copied() {
```

Comment:
We do not want to continue looping once we reach genesis.

```suggestion
                let prev_cached_header = height
                    .checked_sub(1)
                    .and_then(|h| self.headers.get(&h))
                    .copied();
                match prev_cached_header {
```